File: src/test/java/cube8540/oauth/authentication/credentials/oauth/client/infra/ClientAPIExceptionTranslatorTest.java
package cube8540.oauth.authentication.credentials.oauth.client.infra;

import cube8540.oauth.authentication.credentials.oauth.client.domain.ClientAuthorizationException;
import cube8540.oauth.authentication.credentials.oauth.client.domain.ClientInvalidException;
import cube8540.oauth.authentication.credentials.oauth.client.domain.ClientNotFoundException;
import cube8540.oauth.authentication.credentials.oauth.client.domain.ClientRegisterException;
import cube8540.oauth.authentication.error.message.ErrorMessage;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;

@DisplayName("클라이언트 에러 변환 클래스 테스트")
class ClientAPIExceptionTranslatorTest {

    private ClientAPIExceptionTranslator translator;

    @BeforeEach
    void setup() {
        this.translator = new ClientAPIExceptionTranslator();
    }

    @Test
    @DisplayName("ClientAuthorizationException 변환")
    void translateClientAuthorizationException() {
        ClientAuthorizationException e = mock(ClientAuthorizationException.class);

        ResponseEntity<ErrorMessage<Object>> result = translator.translate(e);
        assertEquals(HttpStatus.UNAUTHORIZED, result.getStatusCode());
    }

    @Test
    @DisplayName("ClientInvalidException 변환")
    void translateClientInvalidException() {
        ClientInvalidException e = mock(ClientInvalidException.class);

        ResponseEntity<ErrorMessage<Object>> result = translator.translate(e);
        assertEquals(HttpStatus.BAD_REQUEST, result.getStatusCode());
    }

    @Test
    @DisplayName("ClientNotFoundException 변환")
    void translateClientNotFoundException() {
        ClientNotFoundException e = mock(ClientNotFoundException.class);

        ResponseEntity<ErrorMessage<Object>> result = translator.translate(e);
        assertEquals(HttpStatus.NOT_FOUND, result.getStatusCode());
    }

    @Test
    @DisplayName("ClientRegisterException 변환")
    void translateClientRegisterException() {
        ClientRegisterException e = mock(ClientRegisterException.class);

        ResponseEntity<ErrorMessage<Object>> result = translator.translate(e);
        assertEquals(HttpStatus.BAD_REQUEST, result.getStatusCode());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/client/infra/ClientAPIExceptionTranslatorTest.kt
package cube8540.oauth.authentication.oauth.client.infra

import cube8540.oauth.authentication.oauth.client.domain.ClientAuthorizationException
import cube8540.oauth.authentication.oauth.client.domain.ClientInvalidException
import cube8540.oauth.authentication.oauth.client.domain.ClientNotFoundException
import cube8540.oauth.authentication.oauth.client.domain.ClientRegisterException
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.http.HttpStatus
import java.util.stream.Stream

class ClientAPIExceptionTranslatorTest {

    private val translator = ClientAPIExceptionTranslator()

    @ParameterizedTest
    @MethodSource(value = ["clientAuthorizationException"])
    fun `translate client authorization exception`(exception: ClientAuthorizationException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.FORBIDDEN)
    }

    @ParameterizedTest
    @MethodSource(value = ["clientInvalidException"])
    fun `translate client invalid exception`(exception: ClientInvalidException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    @ParameterizedTest
    @MethodSource(value = ["clientNotFoundException"])
    fun `translate client not found exception`(exception: ClientNotFoundException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.NOT_FOUND)
    }

    @ParameterizedTest
    @MethodSource(value = ["clientRegisterException"])
    fun `translate client register exception`(exception: ClientRegisterException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    fun clientAuthorizationException() = Stream.of(
        Arguments.of(ClientAuthorizationException.invalidPassword("password"))
    )

    fun clientInvalidException() = Stream.of(
        Arguments.of(ClientInvalidException.instance(emptyList()))
    )

    fun clientNotFoundException() = Stream.of(
        Arguments.of(ClientNotFoundException.instance("not found"))
    )

    fun clientRegisterException() = Stream.of(
        Arguments.of(ClientRegisterException.existsIdentifier("exists identifier"))
    )

}